pyro/infer/mcmc/api.py

```diff
@@ -310,9 +313,19 @@ def __init__(self, kernel, num_samples, warmup_steps=None, initial_params=None,
 
     def run(self, *args, **kwargs):
         z_acc = defaultdict(lambda: [[] for _ in range(self.num_chains)])
+        divergences = [None for _ in range(self.num_chains)]
         for sample, chain_id in self.sampler.run(*args, **kwargs):
-            for k, v in sample.items():
-                z_acc[k][chain_id].append(v)
+            if isinstance(sample, dict):
```

Comment:
It will be nice to keep the interface as decoupled from the HMC kernels as possible (I know that we have some HMC specific checks in init, but better to avoid this as much as possible). My suggestion would be to instead do something like:

```python
num_samples = [0] * args.num_chains
for x, chain_id in self.sampler.run(*args, **kwargs):
    if num_samples[chain_id] = self.num_samples:
        self._diagnostics[chain_id] = x
    else:
        for k, v in x.items():
            z_acc[k][chain_id].append(v)
             ...
```

Here, diagnostics itself can be a dict and for HMC/NUTS it might look like `{"divergence_ids" = torch.Tensor([2, 10, 14, ...])}`. 

Then as you suggested `MCMC.diagnostics` property can combine these chain level information stored in `._diagnostics` and return some dict of diagnostic information per sample site.

Reply:
Yup, I expected that you will have a better idea for this issue. And I was right. :D Thanks!